File: frontend/src/app/pages/certification/certification.ts
import {
  Component,
  OnInit,
  signal
} from '@angular/core';

import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';

import { CertificationService } from '../../services/certification.service';
import { Vendor } from '../../services/vendor';


@Component({
  selector: 'app-certification',
  standalone: true,

  imports: [
    CommonModule,
    FormsModule
  ],

  templateUrl: './certification.html',
  styleUrl: './certification.css'
})


export class Certification implements OnInit {


  // ==========================================
  // CERTIFICATION DATA
  // ==========================================

  certifications =
    signal<any[]>([]);


  // ==========================================
  // VENDOR DATA
  // ==========================================

  vendors =
    signal<any[]>([]);


  // ==========================================
  // LOADING
  // ==========================================

  loading =
    signal(false);


  // ==========================================
  // FORM VISIBILITY
  // ==========================================

  showForm =
    signal(false);


  // ==========================================
  // EDIT MODE
  // ==========================================

  editMode =
    signal(false);

  selectedCertificationId:
    number | null = null;


  // ==========================================
  // FORM DATA
  // ==========================================

  selectedVendorId:
    number | null = null;

  certificationName = '';

  certificateNumber = '';

  issuingAuthority = '';

  issueDate = '';

  expiryDate = '';

  status = 'Active';

  notes = '';


  // ==========================================
  // MESSAGES
  // ==========================================

  successMessage = '';

  errorMessage = '';


  // ==========================================
  // CONSTRUCTOR
  // ==========================================

  constructor(

    private certificationService:
      CertificationService,

    private vendorService:
      Vendor

  ) {}


  // ==========================================
  // INITIALIZE
  // ==========================================

  ngOnInit(): void {

    this.loadVendors();

    this.loadCertifications();

  }


  // ==========================================
  // LOAD VENDORS
  // ==========================================

  loadVendors(): void {

    this.vendorService
      .getVendors()
      .subscribe({

        next: (response: any[]) => {

          this.vendors.set(response);

        },

        error: (error: any) => {

          console.error(
            'Failed to load vendors:',
            error
          );

        }

      });

  }


  // ==========================================
  // LOAD CERTIFICATIONS
  // ==========================================

  loadCertifications(): void {

    this.loading.set(true);

    this.certificationService
      .getCertifications()
      .subscribe({

        next: (response: any[]) => {

          this.certifications.set(
            response
          );

          this.loading.set(false);

        },

        error: (error: any) => {

          console.error(
            'Failed to load certifications:',
            error
          );

          this.certifications.set([]);

          this.loading.set(false);

        }

      });

  }


  // ==========================================
  // OPEN ADD FORM
  // ==========================================

  openAddForm(): void {

    this.editMode.set(false);

    this.selectedCertificationId =
      null;

    this.resetForm();

    this.showForm.set(true);

  }


  // ==========================================
  // OPEN EDIT FORM
  // ==========================================

  openEditForm(
    certification: any
  ): void {

    this.editMode.set(true);

    this.selectedCertificationId =
      certification.id;

    this.selectedVendorId =
      certification.vendor_id;

    this.certificationName =
      certification.certification_name;

    this.certificateNumber =
      certification.certificate_number;

    this.issuingAuthority =
      certification.issuing_authority || '';

    this.issueDate =
      certification.issue_date;

    this.expiryDate =
      certification.expiry_date;

    this.status =
      certification.status || 'Active';

    this.notes =
      certification.notes || '';

    this.showForm.set(true);

  }


  // ==========================================
  // CLOSE FORM
  // ==========================================

  closeForm(): void {

    this.showForm.set(false);

    this.resetForm();

  }


  // ==========================================
  // RESET FORM
  // ==========================================

  resetForm(): void {

    this.selectedVendorId =
      null;

    this.certificationName = '';

    this.certificateNumber = '';

    this.issuingAuthority = '';

    this.issueDate = '';

    this.expiryDate = '';

    this.status = 'Active';

    this.notes = '';

    this.selectedCertificationId =
      null;

  }


  // ==========================================
  // SAVE CERTIFICATION
  // ==========================================

  saveCertification(): void {

    this.successMessage = '';

    this.errorMessage = '';


    if (
      !this.selectedVendorId ||
      !this.certificationName ||
      !this.certificateNumber ||
      !this.issueDate ||
      !this.expiryDate
    ) {

      this.errorMessage =
        'Please fill all required fields.';

      return;

    }


    const data = {

      vendor_id:
        this.selectedVendorId,

      certification_name:
        this.certificationName,

      certificate_number:
        this.certificateNumber,

      issuing_authority:
        this.issuingAuthority || null,

      issue_date:
        this.issueDate,

      expiry_date:
        this.expiryDate,

      status:
        this.status,

      notes:
        this.notes || null

    };


    if (
      this.editMode() &&
      this.selectedCertificationId
    ) {

      this.certificationService
        .updateCertification(
          this.selectedCertificationId,
          data
        )
        .subscribe({

          next: () => {

            this.successMessage =
              'Certification updated successfully.';

            this.closeForm();

            this.loadCertifications();

          },

          error: (error: any) => {

            console.error(
              'Failed to update certification:',
              error
            );

            this.errorMessage =
              error?.error?.detail ||
              'Unable to update certification.';

          }

        });

      return;

    }


    this.certificationService
      .createCertification(data)
      .subscribe({

        next: () => {

          this.successMessage =
            'Certification created successfully.';

          this.closeForm();

          this.loadCertifications();

        },

        error: (error: any) => {

          console.error(
            'Failed to create certification:',
            error
          );

          this.errorMessage =
            error?.error?.detail ||
            'Unable to create certification.';

        }

      });

  }


  // ==========================================
  // DELETE CERTIFICATION
  // ==========================================

  deleteCertification(
    id: number
  ): void {

    if (
      !confirm(
        'Are you sure you want to delete this certification?'
      )
    ) {

      return;

    }


    this.certificationService
      .deleteCertification(id)
      .subscribe({

        next: () => {

          this.successMessage =
            'Certification deleted successfully.';

          this.loadCertifications();

        },

        error: (error: any) => {

          console.error(
            'Failed to delete certification:',
            error
          );

          this.errorMessage =
            error?.error?.detail ||
            'Unable to delete certification.';

        }

      });

  }


  // ==========================================
  // CHECK EXPIRY
  // ==========================================

  getExpiryStatus(
    expiryDate: string
  ): string {

    const today =
      new Date();

    const expiry =
      new Date(expiryDate);


    if (expiry < today) {

      return 'Expired';

    }


    const difference =
      expiry.getTime() -
      today.getTime();


    const days =
      Math.ceil(
        difference /
        (1000 * 60 * 60 * 24)
      );


    if (days <= 30) {

      return 'Expiring Soon';

    }


    return 'Active';

  }


  // ==========================================
  // EXPIRING COUNT
  // ==========================================

  getExpiringCount(): number {

    return this.certifications()
      .filter(
        certification =>
          this.getExpiryStatus(
            certification.expiry_date
          ) === 'Expiring Soon'
      )
      .length;

  }


  // ==========================================
  // EXPIRED COUNT
  // ==========================================

  getExpiredCount(): number {

    return this.certifications()
      .filter(
        certification =>
          this.getExpiryStatus(
            certification.expiry_date
          ) === 'Expired'
      )
      .length;

  }

}

File: backend/app/routers/certification.py
from datetime import date, timedelta

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.certification import Certification
from app.models.vendor import Vendor
from app.schemas.certification import CertificationCreate
from app.utils.security import get_current_email


router = APIRouter(
    prefix="/certifications",
    tags=["Certifications"]
)


# ==========================================
# CREATE CERTIFICATION
# ==========================================

@router.post("/")
def create_certification(
    data: CertificationCreate,
    db: Session = Depends(get_db),
    current_email: str = Depends(get_current_email)
):

    # Check vendor

    vendor = db.query(Vendor).filter(
        Vendor.id == data.vendor_id
    ).first()

    if not vendor:

        raise HTTPException(
            status_code=404,
            detail="Vendor not found"
        )


    # Check duplicate certificate number

    existing = db.query(Certification).filter(
        Certification.certificate_number ==
        data.certificate_number
    ).first()

    if existing:

        raise HTTPException(
            status_code=400,
            detail="Certificate number already exists"
        )


    # Check dates

    if data.expiry_date < data.issue_date:

        raise HTTPException(
            status_code=400,
            detail="Expiry date cannot be before issue date"
        )


    certification = Certification(

        vendor_id=data.vendor_id,

        certification_name=data.certification_name,

        certificate_number=data.certificate_number,

        issuing_authority=data.issuing_authority,

        issue_date=data.issue_date,

        expiry_date=data.expiry_date,

        status=data.status,

        notes=data.notes

    )


    db.add(certification)

    db.commit()

    db.refresh(certification)


    return certification


# ==========================================
# GET ALL CERTIFICATIONS
# ==========================================

@router.get("/")
def get_certifications(
    db: Session = Depends(get_db),
    current_email: str = Depends(get_current_email)
):

    certifications = db.query(
        Certification
    ).all()


    return certifications


# ==========================================
# GET SINGLE CERTIFICATION
# ==========================================

@router.get("/{certification_id}")
def get_certification(
    certification_id: int,
    db: Session = Depends(get_db),
    current_email: str = Depends(get_current_email)
):

    certification = db.query(
        Certification
    ).filter(
        Certification.id == certification_id
    ).first()


    if not certification:

        raise HTTPException(
            status_code=404,
            detail="Certification not found"
        )


    return certification


# ==========================================
# GET CERTIFICATIONS BY VENDOR
# ==========================================

@router.get("/vendor/{vendor_id}")
def get_vendor_certifications(
    vendor_id: int,
    db: Session = Depends(get_db),
    current_email: str = Depends(get_current_email)
):

    vendor = db.query(Vendor).filter(
        Vendor.id == vendor_id
    ).first()


    if not vendor:

        raise HTTPException(
            status_code=404,
            detail="Vendor not found"
        )


    certifications = db.query(
        Certification
    ).filter(
        Certification.vendor_id == vendor_id
    ).all()


    return certifications


# ==========================================
# UPDATE CERTIFICATION
# ==========================================

@router.put("/{certification_id}")
def update_certification(
    certification_id: int,
    data: CertificationCreate,
    db: Session = Depends(get_db),
    current_email: str = Depends(get_current_email)
):

    certification = db.query(
        Certification
    ).filter(
        Certification.id == certification_id
    ).first()


    if not certification:

        raise HTTPException(
            status_code=404,
            detail="Certification not found"
        )


    # Check vendor

    vendor = db.query(Vendor).filter(
        Vendor.id == data.vendor_id
    ).first()

    if not vendor:

        raise HTTPException(
            status_code=404,
            detail="Vendor not found"
        )


    # Check duplicate certificate number

    existing = db.query(Certification).filter(
        Certification.certificate_number ==
        data.certificate_number,
        Certification.id != certification_id
    ).first()

    if existing:

        raise HTTPException(
            status_code=400,
            detail="Certificate number already exists"
        )


    # Check dates

    if data.expiry_date < data.issue_date:

        raise HTTPException(
            status_code=400,
            detail="Expiry date cannot be before issue date"
        )


    certification.vendor_id = data.vendor_id

    certification.certification_name = (
        data.certification_name
    )

    certification.certificate_number = (
        data.certificate_number
    )

    certification.issuing_authority = (
        data.issuing_authority
    )

    certification.issue_date = data.issue_date

    certification.expiry_date = data.expiry_date

    certification.status = data.status

    certification.notes = data.notes


    db.commit()

    db.refresh(certification)


    return certification


# ==========================================
# DELETE CERTIFICATION
# ==========================================

@router.delete("/{certification_id}")
def delete_certification(
    certification_id: int,
    db: Session = Depends(get_db),
    current_email: str = Depends(get_current_email)
):

    certification = db.query(
        Certification
    ).filter(
        Certification.id == certification_id
    ).first()


    if not certification:

        raise HTTPException(
            status_code=404,
            detail="Certification not found"
        )


    db.delete(certification)

    db.commit()


    return {
        "message":
        "Certification deleted successfully"
    }


# ==========================================
# EXPIRING CERTIFICATIONS
# ==========================================

@router.get("/alerts/expiring")
def get_expiring_certifications(
    days: int = 30,
    db: Session = Depends(get_db),
    current_email: str = Depends(get_current_email)
):

    today = date.today()

    expiry_limit = today + timedelta(
        days=days
    )


    certifications = db.query(
        Certification
    ).filter(

        Certification.expiry_date >= today,

        Certification.expiry_date <= expiry_limit

    ).all()


    return certifications


# ==========================================
# EXPIRED CERTIFICATIONS
# ==========================================

@router.get("/alerts/expired")
def get_expired_certifications(
    db: Session = Depends(get_db),
    current_email: str = Depends(get_current_email)
):

    today = date.today()


    certifications = db.query(
        Certification
    ).filter(
        Certification.expiry_date < today
    ).all()


    return certifications